### ISSUE TITLE
Allow for batch updates

### ISSUE THREAD

**Issue by puf** [firebase/FirebaseUI-Android#458]
The FirebaseUI adapters surface an atomic event for each update to a child item. In some cases it may be more efficient to update the UI only after a batch up updates has been received.

I'd like it if the FirebaseArray and adapters would raise an additional event when a `value` event is received. Given that 'value' events are guaranteed to fire after the corresponding `child_` events, this would be the perfect way to implement such an event.

    public void onDataChanged()

Note that we don't pass in a DataSnapshot here. The idea is that this is a flow-event only, you're supposed to capture the data in the regular child events and just just `onDataChanged` to capture the completion moment.

**Comment by puf** [comment]
@morganchen12 @samtstern PTAL

**Comment by samtstern** [comment]
I like the idea of the event, but it doesn't really allow for batch updates (which is good, IMO).  It does allow for batch events such as removing loading spinners and checking for emptiness.

**Comment by puf** [comment]
@samtstern That would actually require only a `onReady()` event, which is already covered in #118 . This would be a more generic version of that, firing not just for the initial `value` event, but for each of them.

**Comment by samtstern** [comment]
I like this better than `onReady()` since it's more versatile and matches the name of the real underlying event, but what's something you might do with the second or third event?

**Comment by puf** [comment]
Say for example you used a multi-path update to swap multiple items around.
That would raise multiple `child_moved` events. Having this method would
allow you to capture all these changes and then trigger a single UI update.
I'm not sure if Android supports that, but iOS definitely does. E.g.
http://stackoverflow.com/questions/12656648/uicollectionview-performing-updates-using-performbatchupdates

**Comment by samtstern** [comment]
Yeah with the way things are done right now on Android you couldn't "catch" all of the updates and do a batch.  But I guess you could use it for something like suspending animations and turning them back on during a big change, etc.

Developers will be creative I am sure :-)

**Comment by morganchen12** [comment]
I can't remember what my train of thought was, but on iOS I figured it would be simpler to only observe value events and do the diffing on the client. Maybe I thought it would be simpler for consumers extending the class to not have to worry about internal state aggregation between value events.

Anyway, there is also an issue on iOS https://github.com/firebase/FirebaseUI-iOS/issues/211 that would benefit from only observing value events.

**Comment by puf** [comment]
Yup. That's precisely the type of use-case I had in mind. Many UI
frameworks have a way to mark bulk animations.

I'd really prefer to have the method be data-less though. So it just
demarcates the end of an update. That way we prevent having two paths to
process data, and instead just have mandatory child_ methods and an
additional optional demarcation method in case you have a use-case that
calls for that.

**Comment by morganchen12** [comment]
That also seems reasonable.

Implementation details aside, this is something I'd like to support. Filed on iOS as https://github.com/firebase/FirebaseUI-iOS/issues/222

**Comment by samtstern** [comment]
A fix for this issue has been released in version `1.1.0`
